File: Graph/207. Course Schedule/courseSchedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        # map each course to prerequesites list   
        pre_map = { i:[] for i in range(numCourses) }
        # add prerequsit to the map 
        for course, prerequisite in prerequisites:
            pre_map[course].append(prerequisite)
        
        # print(pre_map)
        
        # all courses along the current DFS path
        visitSet = set()
        # check based on each course, using numCourses
        def dfs(course):
            if course in visitSet:
                return False
            if pre_map[course] == []:
                return True
            
            # add course       
            visitSet.add(course)

            # check pre on each course
            # course (key) -> pre (value)  
            for pre in pre_map[course]:
                # if no loop in the pre, this will not run
                # visit pre 
                if not dfs(pre): return False
            
            # if the for above never runs
            # clear the course in the set, and map
            # note, pre is still in the set
            visitSet.remove(course)
            pre_map[course] = []
            return True
        
        # for the graph that is not fully connected
        # they will not call each other based on the pre
        # 1->2, 3->4
        for course in range(numCourses):
            if not dfs(course): return False
        return True
        

File: Graph/207. Course Schedule/courseSchedule.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
 var canFinish = function(numCourses, prerequisites) {
    let preMap = new Map();
    let visited = new Set(); 
    
    // list every course with corresponding pre (can be empty)
    // 3
    // [[1,0],[2,1]]
    // 0 => [], 1 => [0], 2=> [1]
    for (let i = 0; i < numCourses; i++) {
        preMap.set(i, []);
    }
    for (let [course, pre] of prerequisites) {
        preMap.get(course).push(pre)
    }
    
    
    for (let course = 0; course < numCourses; course++) {
        if (dfs(course) === false)
            return false;
    }
    return true;
    
    function dfs(course) {
        if (visited.has(course)) {
            return false;
        }
        if (preMap.has(course) && preMap.get(course).length === 0) {
            return true;
        }
        
        visited.add(course);
        
        // for ... of (numeric)
        for (const pre of preMap.get(course)) {
            if (dfs(pre) === false)
                return false;
        }
        
        visited.delete(course);
        if (preMap.has(course))
            preMap.get(course).length = 0;
        return true;
    } 
    
};
